File: leetcode/valid_parentheses/valid_parentheses.java
package valid_parentheses;

import java.util.Map;
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        if(s.length() % 2 != 0) return false;

        Map<Character, Character> openByClose = Map.of(
            ')', '(', 
            '}', '{',
            ']', '['
        );

        Stack<Character> bracketsStack = new Stack<>();
        for(Character ch : s.toCharArray()) {
            if(openByClose.containsKey(ch)) {
                if(bracketsStack.empty() || openByClose.get(ch) != bracketsStack.peek()) {
                    return false;
                } else {
                    bracketsStack.pop();
                }
            } else {
                bracketsStack.push(ch);
            }
        }
        return bracketsStack.empty();
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.isValid("()[]{}"));
    }
}

File: leetcode/valid_parentheses/valid_parentheses.cpp
#include<iostream>
#include<unordered_map>
#include<string>
#include<stack>

using std::string;
using std::unordered_map;
using std::stack;


class Solution {
public:
    bool isValid(string s) {
        if(s.length() % 2 != 0) return false;
        unordered_map<char, char> openByClose = {
            {')','('},
            {'}','{'},
            {']','['}
        };

        stack<char> bracketsStack;
        for(char ch : s) {
            if(openByClose.find(ch) != openByClose.end()) {
                if(bracketsStack.empty() || openByClose[ch] != bracketsStack.top()) {
                    return false;
                } 
                bracketsStack.pop();
            } else {
                bracketsStack.push(ch);
            }
        }

        return bracketsStack.empty();
    }
};

int main() {
    Solution s;
    std::cout << s.isValid("([])");
}